### PR TITLE
Pipenets

### PR DESCRIPTION
**Pull request by Aranclanos**
### Purpose
Adds a WIP pipenet system, with two new objects, pipes and connectors.

### Open Questions and Pre-Merge TODOs

- [X]  This fix is tested on the same branch it is PR'ed to.
- [ ]  I correctly commented my code
- [X]  My code is indented with tabs and not spaces
- [X]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [X]  This PR does not bring up any new compile errors
- [X]  This PR has been tested in editor
- [ ]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### Notes:
Known issues:
-pipes get sucked/pushed by environmental air
-cant rotate pipes
-all pipes are north/south
-advanced pipes (vents, scrubbers, connectors) connect both sides even if their sprite is for south
-pipe sprites go over canisters
-canisters have no sprite to indicate the connection to the connector
-canisters can be pushed or pulled when connected
-pipes can be pulled while anchored
-map lacks pipes, vents and scrubbers don't spawn anchored
-no comments
-clients don't update pipe sprites
-there's nothing to check for a shuttle moving/rotating
~~there are some debug variables~~

### CODE REVIEW

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	public List<Pipe> nodes = new List<Pipe>();
+	public bool anchored = false;
+	public Direction direction = Direction.NORTH;
+	public Sprite[] pipeSprites;
+	public SpriteRenderer spriteRenderer;
+
+	public Pipenet pipenet;
+	public float volume = 70;
+
+	public int ARANpipenetmembers;
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs]: lol, why would you add your nickname to field names?

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/AirVent.cs [hunk 2]
```diff
@@ -1,55 +1,51 @@
...
 		{
-			if (metaNode.GasMix.Pressure < MinimumPressure)
-			{
-				metaNode.GasMix = new GasMix(GasMixes.Air);
-				subsystemManager.UpdateAt(metaNode.Position);
-			}
+			GasMix gasMix = pipenet.gasMix;
+			pipenet.gasMix = gasMix / 2;
+			metaNode.GasMix = metaNode.GasMix + gasMix;
+			metaDataLayer.UpdateSystemsAt(Vector3Int.RoundToInt(transform.position));
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/AirVent.cs]: You're using `transform.position` *everywhere*. **Don't!** Instead, cache a reference to RegisterTile in Pipe class and use `registerTile.WorldPositionServer`, they're already rounded and reliable. transform.position is clientside fluff, never rely on it if you have RegisterTile.

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	public string ARANpipenetName;
+	public float ARANpipenetVolume;
+
+	public enum Direction
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs:19]: you can use powerful `Orientation` struct, by the way

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+			break;
+		}
+		if (foundPipenet == null)
+		{
+			foundPipenet = new Pipenet();
+		}
+		foundPipenet.AddPipe(this);
+
+		transform.rotation = new Quaternion();
+		transform.position = Vector3Int.RoundToInt(transform.position);
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs]: Do they still get blown away by wind? You could take advantage of `IsTileSnap` and `isNotPushable` bools, making sure they can't be blown away or nudged (and snap to tile like lockers and players)

**Review by fomalsd — CHANGES_REQUESTED** [review]
Not gonna lie, I lol'd about these named fields pretty hard
Check out other comments, there are a few important points

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/AirVent.cs (hunk 2 above)
Comment (fomalsd) [reply UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/AirVent.cs]: Scr 2019-05-17 в 10 18 12

**Comment by Aranclanos** [comment]
if possible id like to have this merged fix the known issues on different PRs so its more organised and easier to review

File: UnityProject/Assets/Scripts/Objects/Canister.cs [hunk 5]
```diff
@@ -28,7 +30,7 @@ public override bool Interact(GameObject originator, Vector3 position, string ha
 		{
 			if(connector == null)
 			{
-				var foundConnectors = MatrixManager.GetAt<Connector>(position.RoundToInt(), true);
+				var foundConnectors = MatrixManager.GetAt<Connector>(registerTile.PositionServer, true);
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Objects/Canister.cs]: MatrixManager works with World positions only – it iterates through all matrices (And `Matrix` class is for local positions), use `WorldPositionServer` here

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+			if (pipe)
+			{
+				nodes.Add(pipe);
+				pipe.nodes.Add(this);
+				pipe.SpriteChange();
+			}
+		}
+	}
+
+	public List<Vector3Int> GetAdjacentTurfs()
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Meta/Atmospherics/Pipenets/Pipe.cs]: this looks redundant, `MetaUtils.GetNeighbors()` probably does what you need

**Comment by fomalsd** [comment]
other than that it's ok to merge

**Comment by Aranclanos** [comment]
Done

**Review by fomalsd — APPROVED** [review]

**Pull request merged**